src/pages/ThorChainLP/components/AddLiquitity/AddLiquidityInput.tsx

```diff
@@ -315,12 +321,135 @@ export const AddLiquidityInput: React.FC<AddLiquidityInputProps> = ({
+  })
+
+  const assetAccountNumberFilter = useMemo(() => {
+    const poolAccountId =
+      accountIdsByChainId[foundPool?.assetId ? fromAssetId(foundPool.assetId).chainId : '']
+
+    return { assetId: asset?.assetId ?? '', accountId: poolAccountId ?? '' }
+  }, [accountIdsByChainId, foundPool?.assetId, asset?.assetId])
+
+  const assetAccountNumber = useAppSelector(s =>
+    selectAccountNumberByAccountId(s, assetAccountNumberFilter),
+  )
+  const allowanceFromAddress = useMemo(() => {
+    if (!asset) return
+    if (!isToken(fromAssetId(asset.assetId).assetReference)) return
+    const poolAccountId =
+      accountIdsByChainId[foundPool?.assetId ? fromAssetId(foundPool.assetId).chainId : '']
+    if (!poolAccountId) return
+    const supportedEvmChainIds = getSupportedEvmChainIds()
+    if (!supportedEvmChainIds.includes(fromAssetId(asset.assetId).chainId as KnownChainIds)) return
+
+    return fromAccountId(poolAccountId).account
+  }, [accountIdsByChainId, asset, foundPool?.assetId])
+
+  const [approvalTxId, setApprovalTxId] = useState<string | null>(null)
+  const serializedTxIndex = useMemo(() => {
+    const poolAccountId =
+      accountIdsByChainId[foundPool?.assetId ? fromAssetId(foundPool.assetId).chainId : '']
+    if (!(approvalTxId && allowanceFromAddress && poolAccountId)) return ''
+    return serializeTxIndex(poolAccountId, approvalTxId, allowanceFromAddress)
+  }, [accountIdsByChainId, foundPool?.assetId, approvalTxId, allowanceFromAddress])
+
+  // @ts-ignore this is wrongly typed
+  const {
+    mutate,
+    isPending: isApprovalMutationPending,
+    isSuccess: isApprovalMutationSuccess,
+  } = useMutation({
+    ...reactQueries.mutations.approve({
+      assetId: asset?.assetId,
+      spender: inboundAddressData?.router,
+      from: allowanceFromAddress,
+      amount: toBaseUnit(actualAssetCryptoLiquidityAmount, asset?.precision ?? 0),
+      wallet,
+      accountNumber: assetAccountNumber,
+    }),
+    onSuccess: (txId: string) => {
+      setApprovalTxId(txId)
+    },
+  })
+
+  const approvalTx = useAppSelector(gs => selectTxById(gs, serializedTxIndex))
+  const isApprovalTxPending = useMemo(
+    () =>
+      isApprovalMutationPending ||
+      (isApprovalMutationSuccess && approvalTx?.status !== TxStatus.Confirmed),
+    [approvalTx?.status, isApprovalMutationPending, isApprovalMutationSuccess],
+  )
+
+  useEffect(() => {
+    if (!approvalTx) return
+    if (isApprovalTxPending) return
+    ;(async () => {
+      await queryClient.invalidateQueries(
+        reactQueries.common.allowanceCryptoBaseUnit(
+          asset?.assetId,
+          inboundAddressData?.router,
+          allowanceFromAddress,
+        ),
+      )
+    })()
+  }, [
+    allowanceFromAddress,
+    approvalTx,
+    asset?.assetId,
+    inboundAddressData?.router,
+    isApprovalTxPending,
+    queryClient,
+  ])
+
+  const { data: allowanceData, isLoading: isAllowanceDataLoading } = useQuery({
+    refetchInterval: 30_000,
```

Comment:
Just as a paranoia to keep polling in case anything goes wrong, but shouldn't really be needed

Reply:
we also have useIsApprovalNeeded which updates on incoming blocks
`src/components/MultiHopTrade/components/MultiHopTradeConfirm/hooks/useIsApprovalNeeded.tsx`

could be useful to go with this new one, or the existing one across the app

Reply:
Yeah I saw it, unfortunately it is tightly coupled to swapper. If we were to decouple it, we could use it, but a decoupled version is what this query effectively brings as you noted. 

I'd be keen to keep this version which is also theoretically more performant and cleaner (not boilerplate wise necessarily but as in, does things the react-query way) in that:

- it doesn't poll for blocks, but uses the success Tx state from the mutation to set the Txid, and in turn the presence of a confirmed Tx  in store to invalidate the allowance query, meaning there's no JSON-RPC call needed to detect blocks (success Tx status is coming through our ws)
- there's no need to poll *every* block, although we do poll allowance at 30s interval for paranoia, which *could* be avoided here if we wanted to
- using a mutation to dispatch the Tx and set its Txid, then using this to decide when to invalidate keeps us flexible. In this particular case, `onSuccess` sets a Txid which in turns we use for Tx detection, then eventually invalidation, sure, but you could imagine a different scenario where there's no Tx broadcasted by us, but instead we just rely on the polling and don't use the broadcast mutation at all, or perhaps use different heuristics

Happy to move swapper consumptions to this as a follow-up, but won't do in this PR because of the risks associated to diffing a diff domain 🙏🏽